src/canopy/llm/cohere.py

```diff
@@ -0,0 +1,327 @@
+                               model_params: Optional[dict] = None
+                               ) -> dict:
+        raise NotImplementedError("Cohere LLM doesn't support function calling")
+
+    async def aenforced_function_call(self,
+                                      system_prompt: str,
+                                      chat_history: Messages,
+                                      function: Function, *,
+                                      max_tokens: Optional[int] = None,
+                                      model_params: Optional[dict] = None):
+        raise NotImplementedError("Cohere LLM doesn't support function calling")
+
+    async def achat_completion(self,
+                               system_prompt: str,
+                               chat_history: Messages,
+                               context: Optional[Context] = None,
+                               *,
+                               stream: bool = False,
+                               max_generated_tokens: Optional[int] = None,
+                               model_params: Optional[dict] = None,
+                               ) -> Union[ChatResponse,
+                                          Iterable[StreamingChatChunk]]:
+        raise NotImplementedError("Cohere LLM doesn't support async chat completion")
+
+    async def agenerate_queries(self,
+                                messages: Messages,
+                                *,
+                                max_generated_tokens: Optional[int] = None,
+                                model_params: Optional[dict] = None,
+                                ) -> List[Query]:
+        raise NotImplementedError("Cohere LLM doesn't support async query generation")
+
+    def _convert_model_params(self, openai_model_params: dict) -> dict:
+        """
+        Convert Open AI model params to Cohere equivalents.
+
+        Args:
+            openai_model_params: model params passed from client to Canopy API in OpenAI format.
+
+        Returns:
+            Model params used with Cohere Chat API.
+        """  # noqa: E501
+        common_params = [
+            "model",
+            "frequency_penalty",
+            "max_tokens",
+            "presence_penalty",
+            "stream",
+            "temperature",
+        ]
+        converted_model_params = {}
+
+        for param in common_params:
+            if param in openai_model_params:
+                converted_model_params[param] = openai_model_params.pop(param)
+
+        # Scale is -2.0 to 2.0 with OpenAI, but -1.0 to 1.0 with Cohere.
+        if presence_penalty := converted_model_params.get("presence_penalty"):
+            converted_model_params = presence_penalty * 0.5
+
+        unrecognized_keys = set(openai_model_params.keys())
+        default_keys = set(self.default_model_params.keys())
+
+        if unrecognized_keys.difference(default_keys) and not self.ignore_unrecognized_params:
+            raise NotImplementedError(
```

Comment:
Could maybe raise a warning here instead of failing, unless otherwise discussed in this PR already

Reply:
My only concern with leaving it a warning is that we slightly fail in case someone has accidentally misspelled one of the supported parameters. They will get some undesired behavior, without even knowing (unless they check the logs).